### PR TITLE
Enable QNN HTP support for Node

### PR DESCRIPTION
**Pull request by joncamp**
### Description
Add support for using Onnx Runtime with Node

### Motivation and Context
Onnx Runtime supports the QNN HTP, but does not support it for Node.js. This adds baseline support for the Onnx Runtime to be used with Node.

Note it does not update the node packages that are distributed officially. This simply patches the onnxruntime.dll to allow 'qnn' to be used as an execution provider.

Testing was done using the existing onnxruntime-node package. The `onnxruntime.dll` and `onnxruntime_binding.node` were swapped into `node_modules\onnxruntime-node\bin\napi-v3\win32\arm64` with the newly built version, then the various QNN dlls and .so files were placed next to the onnxruntime.dll. Testing was performed on a variety of models and applications, but the easiest test is to modify the [node quickstart example](https://github.com/microsoft/onnxruntime-inference-examples/tree/main/js/quick-start_onnxruntime-node).

### CODE REVIEW

**Comment by joncamp** [comment]
@microsoft-github-policy-service agree company="Cephable"


Thanks,

Jon



Jon Campbell

Director of Engineering

mobile: 425-256-0644

email: ***@***.******@***.***>

web: cephable.com<http://cephable.com/>

Book a Meeting With ***@***.***?anonymous&ep=plink>



________________________________

**Comment by fs-eire** [comment]
I have a few questions regarding this QNN HTP feature:

- Is this feature only works ( or it's only planned to be supported ) on Windows/arm64 ? Will other OS/CPU arch also need this feature ?
- Is QNN EP using static link (ie. included in onnxruntime.dll, like DML) or dynamic link (ie. included in onnxruntime_provider_xxx.dll, like CUDA) ?
- Is it compatible with CPU/DML if a build enabled them all? Specifically:
  - If it is running on a non-Qualcomm device (not sure if this assumption exists, please let me know if I made wrong assumption) or an environment where QNN is not supported, is it still be able to load and run other EP?
  - Is the binary works with other non-CPU EP, like DML? (we know that CUDA has some issue that it does not work with DML in one build)
- Do you want to include QNN HTP support in onnxruntime-node by default? If so-
  - Do we already have a build pipeline for release artifacts in the "Zip-*" pipeline?

**Comment by jywu-msft** [comment]
> I have a few questions regarding this QNN HTP feature:
> 
> * Is this feature only works ( or it's only planned to be supported ) on Windows/arm64 ? Will other OS/CPU arch also need this feature ?
> * Is QNN EP using static link (ie. included in onnxruntime.dll, like DML) or dynamic link (ie. included in onnxruntime_provider_xxx.dll, like CUDA) ?
> * Is it compatible with CPU/DML if a build enabled them all? Specifically:
>   
>   * If it is running on a non-Qualcomm device (not sure if this assumption exists, please let me know if I made wrong assumption) or an environment where QNN is not supported, is it still be able to load and run other EP?
>   * Is the binary works with other non-CPU EP, like DML? (we know that CUDA has some issue that it does not work with DML in one build)
> * Do you want to include QNN HTP support in onnxruntime-node by default? If so-
>   
>   * Do we already have a build pipeline for release artifacts in the "Zip-*" pipeline?

i can answer some of these questions.
1) win/arm64 for now.  maybe other platforms/later need to see what those are. (QNN iteself runs on win/arm64, win/x64, linux/x64, android)
2) QNN EP is statically linked to onnxruntime.dll  (that wouldn't change anytime soon)
3) QNN will not run on non-qualcomm HW
4) compatible with CPU.  DML it should be. (but not sure how extensively it's been tested)
5) I think this PR just enables build from source, but it would be nice to eventually support this in a more official manner (pipelines, default options etc.)

**Comment by hans00** [comment]
> * QNN will not run on non-qualcomm HW

QNN CPU might be ok
but as my test, performance will worse then XNNPACK + CPU EP

P.S. QNN CPU seems wrapper of XNNPACK

**Comment by jywu-msft** [comment]
@joncamp fyi, https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=1371836&view=logs&j=90af55dc-07cc-5abc-02f4-1cf38a060872&t=7ee52c27-1516-5118-a868-3d2d34beb196

lib/wasm/session-options.ts:121:31 - error TS2339: Property 'preferredLayout' does not exist on type 'QnnExecutionProviderOption'.

121               if (qnnOptions?.preferredLayout) {
                                  ~~~~~~~~~~~~~~~


Found 1 error in lib/wasm/session-options.ts:121

File: js/node/src/session_options_helper.cc [hunk 1]
```diff
@@ -21,6 +21,9 @@
 #ifdef USE_COREML
 #include "core/providers/coreml/coreml_provider_factory.h"
 #endif
+#ifdef USE_QNN
+//#include "core/providers/qnn/qnn_provider_factory.h"
```
Comment (HectorSVC) [inline js/node/src/session_options_helper.cc:25]: this is not required, we use the general one AppendExecutionProvider which is in API header file.

File: js/node/src/session_options_helper.cc [hunk 2]
```diff
@@ -80,6 +83,13 @@
 #ifdef USE_COREML
     } else if (name == "coreml") {
       Ort::ThrowOnError(OrtSessionOptionsAppendExecutionProvider_CoreML(sessionOptions, coreMlFlags));
+#endif
+#ifdef USE_QNN
+    } else if (name == "qnn") {
+      std::unordered_map<std::string, std::string> qnn_options;
+      qnn_options["backend_path"] = "QnnHtp.dll";
+      qnn_options["enable_htp_fp16_precision"] = "1";
```
[the remark this answers is not in the document]
Comment (jywu-msft) [reply js/node/src/session_options_helper.cc:91]: i think PR is just adding the barebones to make the build work with fp16. we can help fill in the rest.

File: js/web/lib/wasm/session-options.ts [hunk 3]
```diff
@@ -114,6 +114,12 @@ const setExecutionProviders =
               }
             }
             break;
+          case 'qnn':
+            epName = 'QNN';
+            // if (typeof ep !== 'string') {
+            //   const qnnOptions = ep as InferenceSession.QnnExecutionProviderOption;
+            // }
```
Comment (fs-eire) [inline js/web/lib/wasm/session-options.ts:121]: Is this a TODO items that you want to update in seperate PR?

File: js/web/script/test-runner-cli.ts [hunk 4]
```diff
@@ -167,6 +167,7 @@ async function main() {
       cpuOptions: args.cpuOptions,
       webglOptions: args.webglOptions,
       webnnOptions: args.webnnOptions,
+      qnnOptions: args.qnnOptions,
```
Comment (fs-eire) [inline js/web/script/test-runner-cli.ts:170]: no need to put QNN options in ort-web tests as it is not going to support QNN in browser. should remove all changes safely under folder /js/web/

File: js/node/src/session_options_helper.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (joncamp) [reply js/node/src/session_options_helper.cc:91]: That's correct. This is meant to be the barebones requirements to get folks like us unblocked. I hope for other improvements after this one is merged.

File: js/node/src/session_options_helper.cc (hunk 1 above)
Comment (joncamp) [reply js/node/src/session_options_helper.cc]: Removed per request

File: js/web/script/test-runner-cli.ts (hunk 4 above)
Comment (joncamp) [reply js/web/script/test-runner-cli.ts]: removed changes under /js/web per request

**Review by jywu-msft — APPROVED** [review]

**Pull request merged**